domain/blockdevice/service/service.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+	"github.com/juju/juju/domain/blockdevice"
+	"github.com/juju/juju/domain/blockdevice/state"
+)
+
+// State defines an interface for interacting with the underlying state.
+type State interface {
+	BlockDevices(ctx context.Context, machine string) ([]blockdevice.BlockDevice, error)
+	GetMachineInfo(ctx context.Context, machine string) (string, int, error)
+}
+
+// Logger facilitates emitting log messages.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// WatcherFactory describes methods for creating watchers.
+type WatcherFactory interface {
+	NewValuePredicateWatcher(
+		namespace, changeValue string,
+		changeMask changestream.ChangeType,
+		predicate eventsource.Predicate,
+	) (watcher.NotifyWatcher, error)
+}
+
+// Service defines a service for interacting with the underlying state.
+type Service struct {
+	st             State
+	watcherFactory WatcherFactory
+	logger         Logger
+}
+
+// NewService returns a new Service for interacting with the underlying state.
+func NewService(st State, wf WatcherFactory, logger Logger) *Service {
+	return &Service{
+		st:             st,
+		watcherFactory: wf,
+		logger:         logger,
+	}
+}
+
+// WatchBlockDevices returns a new NotifyWatcher watching for
+// changes to block devices associated with the specified machine.
+func (s *Service) WatchBlockDevices(
+	ctx context.Context,
+	machine string,
+) (watcher.NotifyWatcher, error) {
+	machineUUID, life, err := s.st.GetMachineInfo(ctx, machine)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if life == state.Dead {
+		return nil, errors.Errorf("cannot watch block devices on dead machine %q", machine)
+	}
+
+	predicate := func(ctx context.Context, db database.TxnRunner, changes []changestream.ChangeEvent) (bool, error) {
```

Comment:
I wanted to add the logging - hence gathering the changes - because in the past issues have been very had to track down and adding logging like this at the point of consumption from the core infrastructure into the business logic greatly aided in debugging. I am happy to simplify things and remove the logging for now and we can add something later if the need arises.